### PR TITLE
Fix partition estimation with TWCS tables during streaming

### PR DESCRIPTION
**Pull request by raphaelsc**
TWCS tables require partition estimation adjustment as incoming streaming data can be segregated into the time windows. 

Turns out we had two problems in this area that leads to suboptimal bloom filters.

1) With off-strategy enabled, data segregation is postponed, but partition estimation was adjusted as if segregation wasn't postponed. Solved by not adjusting estimation if segregation is postponed.
2) With off-strategy disabled, data segregation is not postponed, but streaming didn't feed any metadata into partition estimation procedure, meaning it had to assume the max windows input data can be segregated into (100). Solved by using schema's default TTL for a precise estimation of window count.

For the future, we want to dynamically size filters (see https://github.com/scylladb/scylladb/issues/2024), especially for TWCS that might have SSTables that are left uncompacted until they're fully expired, meaning that the system won't heal itself in a timely manner through compaction on a SSTable that had partition estimation really wrong.

Fixes https://github.com/scylladb/scylladb/issues/15704.

### CODE REVIEW

**Comment by raphaelsc** [comment]
> * test_replace_node_same_ip_take_write

That's https://github.com/scylladb/scylladb/issues/15713

**Review by denesb — APPROVED** [review]

**Comment by raphaelsc** [comment]
@scylladb/scylla-maint ping for merge.

**Comment by raphaelsc** [comment]
v2: 
Fix condition for estimation using min and max metadata
Improve test to stress the problem fixed in this version

**Comment by raphaelsc** [comment]
> #### 🟢 CI State: SUCCESS
> ✅ - Build ✅ - Sanity Tests ✅ - Unit Tests
> 
> #### Build Details:
> * Build URL: https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4564/
> * Duration: 2 hr 20 min
> * Builder: spider3.cloudius-systems.com

@denesb please consider merging.

File: compaction/time_window_compaction_strategy.cc [hunk 1]
```diff
@@ -184,16 +184,23 @@ class classify_by_timestamp {
...
+uint64_t time_window_compaction_strategy::adjust_partition_estimate(const mutation_source_metadata& ms_meta, uint64_t partition_estimate, schema_ptr s) const {
+    // If not enough information, we assume the worst
+    auto estimated_window_count = max_data_segregation_window_count;
+    auto default_ttl = std::chrono::duration_cast<std::chrono::seconds>(s->default_time_to_live());
+    bool min_and_max_ts_available = ms_meta.min_timestamp && ms_meta.max_timestamp;
+
+    if (!min_and_max_ts_available && default_ttl.count()) {
+        std::chrono::seconds window_size = _options.get_sstable_window_size();
+
+        estimated_window_count = (default_ttl.count() + window_size.count() - 1) / window_size.count();
```
Comment (avikivity) [inline compaction/time_window_compaction_strategy.cc:196]: It's safer to divide duration by duration. I see you divide seconds.count by seconds.count, but if you write it as (default_ttl + window_size - 1s) / window_size, the units cancel out by themselves.

File: compaction/time_window_compaction_strategy.cc [hunk 2]
```diff
@@ -184,16 +184,23 @@ class classify_by_timestamp {
...
+    } else if (min_and_max_ts_available) {
+        auto min_window = get_window_for(_options, *ms_meta.min_timestamp);
+        auto max_window = get_window_for(_options, *ms_meta.max_timestamp);
+        const auto window_size = get_window_size(_options);
+
+        estimated_window_count = (max_window + (window_size - 1) - min_window) / window_size;
```
Comment (avikivity) [inline compaction/time_window_compaction_strategy.cc:202]: like here

File: compaction/time_window_compaction_strategy.cc (hunk 1 above)
Comment (avikivity) [reply compaction/time_window_compaction_strategy.cc:196]: you can simplify the computation: compute the time range that the data spans (either from min_max_timestamp or from the default ttl) and apply the same calculation to the result.

File: compaction/time_window_compaction_strategy.cc [hunk 3]
```diff
@@ -184,16 +184,23 @@ class classify_by_timestamp {
...
     }
-    const auto min_window = get_window_for(_options, *ms_meta.min_timestamp);
-    const auto max_window = get_window_for(_options, *ms_meta.max_timestamp);
-    const auto window_size = get_window_size(_options);
-
-    auto estimated_window_count = (max_window + (window_size - 1) - min_window) / window_size;
 
     return partition_estimate / std::max(1UL, uint64_t(estimated_window_count));
```
Comment (avikivity) [inline compaction/time_window_compaction_strategy.cc:205]: Let's add protection for division by zero which we may have introduced here.

Comment (raphaelsc) [reply compaction/time_window_compaction_strategy.cc:205]: That requires window size to be 0, and it would be a preexisting problem.

Comment (raphaelsc) [reply compaction/time_window_compaction_strategy.cc:205]: We're adding protection against bad settings, refusing creation or altering with them.

**Comment by raphaelsc** [comment]
v3: - reuse existing method to estimate window count with default ttl.

**Comment by mykaul** [comment]
@tgrabiec / @nyh  / @avikivity  - can you please review so we can merge this if approved?

**Pull request merged**